packages/flutter/lib/src/widgets/form.dart

```diff
@@ -139,13 +159,16 @@ class Form extends StatefulWidget {
 /// Typically obtained via [Form.of].
 class FormState extends State<Form> {
   int _generation = 0;
+  bool _hasInteractedByUser = false;
   final Set<FormFieldState<dynamic>> _fields = <FormFieldState<dynamic>>{};
 
   // Called when a form field has changed. This will cause all form fields
   // to rebuild, useful if form fields have interdependencies.
   void _fieldDidChange() {
     if (widget.onChanged != null)
       widget.onChanged();
+    _hasInteractedByUser = _fields
+        .any((FormFieldState<dynamic> field) => field._hasInteractedByUser);
```

Comment:
Could there be any performance concern with this? Do we ever have forms with many many fields?  @chunhtai 

This is probably ok, I just wanted to flag it.  I imagine it would be possible to check only the field that has changed here, but it might not be worth it.

Reply:
It does exactly this: it checks every element in the list and stops if it found that `_hasInteractedByUser ` is true or else keep checking until the last element.

Reply:
One option would be to make the FormFields to update a property in the Form, this would prevent to iterate the whole list.

Reply:
This might be a problem when there are a lot of fields, since it will be called on every key pressed.

Reply:
What are the cases where _hasInteractedByUser gets set to false?  I guess when reset is called, but when else?  Maybe we could make an optimization like this:

```dart
if (!_hasInteractedByUser) {
  _hasInteractedByUser = _fields
        .any((FormFieldState<dynamic> field) => field._hasInteractedByUser);
}
```